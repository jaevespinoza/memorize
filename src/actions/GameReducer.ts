// src/features/gameSlice.js
import { createSlice } from "@reduxjs/toolkit";
import { ISelectedImage } from "../interfaces/AppActionsInterface";

export interface IGameState {
  selectedCards: ISelectedImage[];
  foundCards: ISelectedImage[];
  matchPopup: boolean;
  matchedCard: boolean;
  successfulGames: number;
  success: number;
  errors: number;
  name: string;
}

const gameSlice = createSlice({
  name: "gameApplication",
  initialState: {
    selectedCards: [] as ISelectedImage[],
    foundCards: [] as ISelectedImage[],
    success: 0,
    matchPopup: false,
    matchedCard: false,
    errors: 0,
    name: "",
    successfulGames: 0,
  } as IGameState,
  reducers: {
    /**
     * Sets whether the cards match or not
     */
    setMatchedCard: (state, action) => {
      state.matchedCard = action.payload;
    },
    /**
     * Sets whether the match alert should be shown or not
     */
    setMatchedPopup: (state, action) => {
      state.matchPopup = action.payload;
    },
    /**
     * Sets the selected cards into the state and then checks
     * whether they match or not
     */
    setSelectedCards: (state, action) => {
      if (!state.selectedCards.some((item) => item.id === action.payload.id)) {
        state.selectedCards.push(action.payload);
        checkSelected(state);
      }
    },
    /**
     * Sets the found cards into the state so they are still shown
     */
    setFoundCards: (state, action) => {
      state.foundCards.push(action.payload);
    },
    /**
     * Sets the number of successful matches
     */
    setSuccess: (state, action) => {
      state.success = action.payload;
    },
    /**
     * Sets the number of wrong matches
     */
    setErrors: (state, action) => {
      state.errors = action.payload;
    },
    /**
     * Sets the name of the user into the state
     */
    setName: (state, action) => {
      state.name = action.payload;
    },
    /**
     * Resets the selected cards and then adds to the error
     */
    resetCards: (state) => {
      state.selectedCards = [];
    },
    resetGame: (state) => {
      state.foundCards = [];
      state.selectedCards = [];
      state.success = 0;
      state.errors = 0;
      state.successfulGames += 1;
    },
  },
});

/**
 * Checks the cards that were selected.
 * If the cards match, they are added to the pool of found cards and
 * the success popup shows. If not, they are discarded and the
 * error popup shows.
 * @param state The Application state
 * @returns The updated state with the amount of errors and successes updated
 */
const checkSelected = (state: IGameState) => {
  // If the user hasn't selected more than two cards, we return
  if (state.selectedCards.length < 2) {
    return;
  }
  // We compare the names of the cards
  const checkSelectedCards =
    state.selectedCards[0].name === state.selectedCards[1].name;
  state.matchPopup = true;
  // If they match, we set the matchedCard variable to true. If not, it's false.
  if (checkSelectedCards) {
    state.matchedCard = true;
    successOnMatch(state);
  } else {
    state.matchedCard = false;
    state.errors += 1;
  }
};

const successOnMatch = (state: IGameState) => {
  state.foundCards.push(state.selectedCards[0]);
  state.foundCards.push(state.selectedCards[1]);
  state.selectedCards = [];
  state.success += 1;
};

export const {
  setSelectedCards,
  setFoundCards,
  setMatchedCard,
  resetCards,
  resetGame,
  setMatchedPopup,
  setSuccess,
  setErrors,
  setName,
} = gameSlice.actions;
export default gameSlice.reducer;
