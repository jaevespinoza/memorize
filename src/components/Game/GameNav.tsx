import { useSelector } from "react-redux";
import "./styles.scss";
import { RootState } from "../../../config/store";

/**
 * Nav of the site that shows the name of the game, and the
 * counter for both the successes and the errors.
 * It uses selectors to access the state and show the values.
 */
const GameNav = () => {
  const successCount = useSelector((state: RootState) => state.game.success);
  const errorCount = useSelector((state: RootState) => state.game.errors);
  return (
    <nav className="navbar navbar-dark bg-primary game-container__header fixed-top">
      <div className="container">
        <div className="d-flex justify-content-center align-items-center flex-grow-1">
          <h3 className="game-container__header__title mb-0">Memorize!</h3>
        </div>
        <div className="game-container__header__counters text-white">
          <span className="badge badge-pill badge-success">
            Success: {successCount}
          </span>
          <span className="badge badge-pill badge-danger">
            Errors: {errorCount}
          </span>
        </div>
      </div>
    </nav>
  );
};

export default GameNav;
