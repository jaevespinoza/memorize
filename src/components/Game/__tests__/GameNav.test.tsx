import { configureStore } from "@reduxjs/toolkit";
import { ToolkitStore } from "@reduxjs/toolkit/dist/configureStore";
import { beforeEach, describe, expect, test } from "vitest";
import GameReducer, {
  IGameState,
  setErrors,
  setSuccess,
} from "../../../actions/GameReducer";
import { render, screen } from "@testing-library/react";
import { Provider } from "react-redux";
import GameNav from "../GameNav";

describe("Game Nav", () => {
  let store: ToolkitStore<{ game: IGameState }>;

  beforeEach(() => {
    store = configureStore({ reducer: { game: GameReducer } });
  });

  test("Renders properly", () => {
    const nav = render(
      <Provider store={store}>
        <GameNav />
      </Provider>
    );

    expect(nav).toMatchSnapshot();
  });

  test("Shows amount of success", () => {
    store.dispatch(setSuccess(2));
    render(
      <Provider store={store}>
        <GameNav />
      </Provider>
    );

    expect(screen.getByTestId("success-count").textContent).toBe("Success: 2");
  });

  test("Shows amount of errors", () => {
    store.dispatch(setErrors(2));
    render(
      <Provider store={store}>
        <GameNav />
      </Provider>
    );

    expect(screen.getByTestId("error-count").textContent).toBe("Errors: 2");
  });

  test("Shows title of page", () => {
    render(
      <Provider store={store}>
        <GameNav />
      </Provider>
    );

    expect(screen.getByTestId("header-title").textContent).toBe("Memorize!");
  });
});
