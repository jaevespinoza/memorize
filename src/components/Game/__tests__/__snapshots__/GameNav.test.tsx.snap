// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`Game Nav > Renders properly 1`] = `
{
  "asFragment": [Function],
  "baseElement": <body>
    <div>
      <nav
        class="navbar navbar-dark bg-primary game-container__header fixed-top"
      >
        <div
          class="container"
        >
          <div
            class="d-flex justify-content-center align-items-center flex-grow-1"
          >
            <h3
              class="game-container__header__title mb-0"
              data-testid="header-title"
            >
              Memorize!
            </h3>
          </div>
          <div
            class="game-container__header__counters text-white"
          >
            <span
              class="badge badge-pill badge-success"
              data-testid="success-count"
            >
              Success: 
              0
            </span>
            <span
              class="badge badge-pill badge-danger"
              data-testid="error-count"
            >
              Errors: 
              0
            </span>
          </div>
        </div>
      </nav>
    </div>
  </body>,
  "container": <div>
    <nav
      class="navbar navbar-dark bg-primary game-container__header fixed-top"
    >
      <div
        class="container"
      >
        <div
          class="d-flex justify-content-center align-items-center flex-grow-1"
        >
          <h3
            class="game-container__header__title mb-0"
            data-testid="header-title"
          >
            Memorize!
          </h3>
        </div>
        <div
          class="game-container__header__counters text-white"
        >
          <span
            class="badge badge-pill badge-success"
            data-testid="success-count"
          >
            Success: 
            0
          </span>
          <span
            class="badge badge-pill badge-danger"
            data-testid="error-count"
          >
            Errors: 
            0
          </span>
        </div>
      </div>
    </nav>
  </div>,
  "debug": [Function],
  "findAllByAltText": [Function],
  "findAllByDisplayValue": [Function],
  "findAllByLabelText": [Function],
  "findAllByPlaceholderText": [Function],
  "findAllByRole": [Function],
  "findAllByTestId": [Function],
  "findAllByText": [Function],
  "findAllByTitle": [Function],
  "findByAltText": [Function],
  "findByDisplayValue": [Function],
  "findByLabelText": [Function],
  "findByPlaceholderText": [Function],
  "findByRole": [Function],
  "findByTestId": [Function],
  "findByText": [Function],
  "findByTitle": [Function],
  "getAllByAltText": [Function],
  "getAllByDisplayValue": [Function],
  "getAllByLabelText": [Function],
  "getAllByPlaceholderText": [Function],
  "getAllByRole": [Function],
  "getAllByTestId": [Function],
  "getAllByText": [Function],
  "getAllByTitle": [Function],
  "getByAltText": [Function],
  "getByDisplayValue": [Function],
  "getByLabelText": [Function],
  "getByPlaceholderText": [Function],
  "getByRole": [Function],
  "getByTestId": [Function],
  "getByText": [Function],
  "getByTitle": [Function],
  "queryAllByAltText": [Function],
  "queryAllByDisplayValue": [Function],
  "queryAllByLabelText": [Function],
  "queryAllByPlaceholderText": [Function],
  "queryAllByRole": [Function],
  "queryAllByTestId": [Function],
  "queryAllByText": [Function],
  "queryAllByTitle": [Function],
  "queryByAltText": [Function],
  "queryByDisplayValue": [Function],
  "queryByLabelText": [Function],
  "queryByPlaceholderText": [Function],
  "queryByRole": [Function],
  "queryByTestId": [Function],
  "queryByText": [Function],
  "queryByTitle": [Function],
  "rerender": [Function],
  "unmount": [Function],
}
`;
